### PR TITLE
Separate single- and multi-group radiation modules

### PR DESCRIPTION
**Pull request by chongchonghe**
### Description

The single-group radiation scheme has complex high-order terms, while the multi-group scheme has complexy opacity functions and Jacobian matrix. Previously, these two schemes are mixed together. It's a mess. In this PR, I splitted `AddSourceTerms` into `AddSourceTermsSingleGroup` and `AddSourceTermsMultiGroup`. In `AddSourceTermsSingleGroup`, whether or not the dust model is enabled, the Jacobian is a 2x2 matrix and can be inverted with a simple analytical expression. Second or higher order terms are supported. In `AddSourceTermsMultiGroup`, complexy multigroup opacity models are included but only 0th and 1st order terms are supported. The Jacobian is a (Ng + 1) by (Ng + 1) sparse matrix. 

### Related issues

None.

### Checklist

_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_

- [x] I have added a description (see above).
- [x] I have added a link to any related issues see (see above).
- [x] I have read the [[Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md)](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [ ] I have added tests for any new physics that this PR adds to the code.
- [x] I have tested this PR on my local computer and all tests pass.
- [x] I have manually triggered the GPU tests with the magic comment `/azp run`.
- [x] I have requested a reviewer for this PR.

### CODE REVIEW

**Comment by chongchonghe** [comment]
@BenWibking Both GPU tests failed due to `disk quota exceeded`. Everything else should be fine.

File: src/problems/RadDust/test_rad_dust.cpp
Comment (markkrumholz) [inline src/problems/RadDust/test_rad_dust.cpp:80]: The notation here is confusing. Is temperature really T^4? If so, please rename the variable to that. If not, can you clarify why the source term is only radiation_constant * T and not T^4?

File: src/problems/RadDust/test_rad_dust.cpp
Comment (markkrumholz) [inline src/problems/RadDust/test_rad_dust.cpp:86]: Same question as previous comment. I presume the reason it's just radiation constant is that you're taking the derivative with respect to T^4 and not T, but the variable naming and comments should make that clear.

**Review by markkrumholz — COMMENTED** [review]
I left a couple of comments, and then there some unused variable warnings, but once those are resolved this looks fine and I can approve.

File: src/problems/RadDust/test_rad_dust.cpp
Comment (chongchonghe) [reply src/problems/RadDust/test_rad_dust.cpp:80]: I have to assume B = a_rad * T_d in this test in order to linearize this problem so that I can obtain an analytic solution. I added a comment to explain this. The motivation of doing this is also explained before Eq. 44 of the overleaf document quokka-thermochemistry.

**Comment by chongchonghe** [comment]
> I left a couple of comments, and then there some unused variable warnings, but once those are resolved this looks fine and I can approve.

I had to assume B(T) = a_rad * T in that test in order to obtain an analytic solution. 

The unused variable warning are some false alarms and some unavoidable. They are used in several `if constexpr`, and if you compile the code with if constexpr false, it will think the variables are unused. I tried to use `amrex::ignore_unused` to suppress some of them.

File: src/radiation/radiation_system.hpp
Comment (BenWibking) [inline src/radiation/radiation_system.hpp]: Is there a reason to keep this `else` statement?

File: src/radiation/radiation_system.hpp
Comment (BenWibking) [inline src/radiation/radiation_system.hpp]: The source term implementation has become very complicated. Maybe this function can go in a separate header file?

File: src/radiation/radiation_system.hpp
Comment (BenWibking) [inline src/radiation/radiation_system.hpp]: The multigroup source term implementation has become very complicated. Maybe this can go in a separate header file?

**Comment by BenWibking** [comment]
> The unused variable warning are some false alarms and some unavoidable. They are used in several `if constexpr`, and if you compile the code with if constexpr false, it will think the variables are unused. I tried to use `amrex::ignore_unused` to suppress some of them.

Can you suppress them all with `amrex::ignore_unused`? Or is there an issue you are running into if you try to do this?

File: src/radiation/radiation_system.hpp
Comment (chongchonghe) [reply src/radiation/radiation_system.hpp]: No. I have deleted it.

**Comment by chongchonghe** [comment]
> > The unused variable warning are some false alarms and some unavoidable. They are used in several `if constexpr`, and if you compile the code with if constexpr false, it will think the variables are unused. I tried to use `amrex::ignore_unused` to suppress some of them.
> 
> Can you suppress them all with `amrex::ignore_unused`? Or is there an issue you are running into if you try to do this?

I already have `amrex::ignore_unused` for all of them, but the warning still show up. It might be a bug of github-advanced-security. I'll ignore them for now.

File: src/radiation/radiation_system.hpp
Comment (chongchonghe) [reply src/radiation/radiation_system.hpp]: Maybe in a future PR?

File: src/radiation/radiation_system.hpp
Comment (chongchonghe) [reply src/radiation/radiation_system.hpp]: This is a good point. Maybe in a future PR?

**Comment by BenWibking** [comment]
> > > The unused variable warning are some false alarms and some unavoidable. They are used in several `if constexpr`, and if you compile the code with if constexpr false, it will think the variables are unused. I tried to use `amrex::ignore_unused` to suppress some of them.
> > 
> > 
> > Can you suppress them all with `amrex::ignore_unused`? Or is there an issue you are running into if you try to do this?
> 
> I already have `amrex::ignore_unused` for all of them, but the warning still show up. It might be a bug of github-advanced-security. I'll ignore them for now.

Ok. You can dismiss them as false positives using the drop-down menu.

File: src/radiation/radiation_system.hpp
Comment (BenWibking) [reply src/radiation/radiation_system.hpp]: Is there a reason not to do it in this PR? It should take about 60 seconds to implement.

Comment (BenWibking) [reply src/radiation/radiation_system.hpp]: In this case, the right way to do it is probably with a recursive include. That is, at the end of `radiation_system.hpp`, you can do
```
#include "AddSourceTerms.hpp  // IWYU pragma: export
#include "AddSourceTermsMultigroup.hpp  // IWYU pragma: export
```
and then in each of those header files, just cut-and-paste the functions, and add this at the top of each file:
```
// IWYU pragma: private; include "radiation_system.h"
```
(See https://clangd.llvm.org/guides/include-cleaner#iwyu-pragmas)

Comment (chongchonghe) [reply src/radiation/radiation_system.hpp]: OK, this looks good. I've updated it. The code compiles and runs without any problem. The only drawback I'm experiencing is that vscode won't do variable indexing inside add_source_terms_multi_group.hpp, but I think this can be fixed.

**Comment by chongchonghe** [comment]
@BenWibking Can you approve this? I'll merge it once the GPU tests finish.

**Review by BenWibking — DISMISSED** [review]

**Comment by chongchonghe** [comment]
I'll hold this for a little bit. I made some change to the code structure and fixed the clangd/IntelliSense problem.

**Comment by chongchonghe** [comment]
This is an alternative approach I showed in the meeting today. This way the auto-complete and syntax highlighting works fine in VSCode. https://github.com/quokka-astro/quokka/tree/chong/separate-v2

**Comment by BenWibking** [comment]
> This is an alternative approach I showed in the meeting today. This way the auto-complete and syntax highlighting works fine in VSCode. https://github.com/quokka-astro/quokka/tree/chong/separate-v2

Can you see if it works with a recursive include, e.g.:

radiation_system.hpp
```
// end of file
#include "source_terms.hpp" // IWYU: blahblah
```

source_terms.hpp
```
//start of file
// IWYU: blahblah1
#include "radiation_system.hpp" // IWYU: blahblah2
```

Since there are include guards for `radiation_system.hpp`, it should not actually cause an infinite loop, but it might be enough for VSCode to figure out where the symbols are coming from.

**Comment by BenWibking** [comment]
This approach should be the same as "method 3" here, except that their version of `source_terms.hpp` (incorrectly) ends in `.cpp`: https://www.codeproject.com/Articles/48575/How-to-Define-a-Template-Class-in-a-h-File-and-Imp

If this still doesn't work, it should be reported as a bug in `clangd`: https://github.com/clangd/clangd/issues.

**Comment by chongchonghe** [comment]
@BenWibking Works fine now.

**Review by BenWibking — APPROVED** [review]

**Pull request merged**
